cmd/agent-smoke/internal/smoke.go

```diff
@@ -0,0 +1,105 @@
+package smoke
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/oklog/run"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// Smoke is the top level object for a smoke test.
+type Smoke struct {
+	logger log.Logger
+	tasks  []repeatingTask
+}
+
+// Options struct to pass configuration to the Smoke constructor.
+type Options struct {
+	Kubeconfig        string
+	Namespace         string
+	ChaosFrequency    time.Duration
+	MutationFrequency time.Duration
+}
+
+// NewSmokeTest is the constructor for a Smoke object.
+func NewSmokeTest(logger log.Logger, opts Options) (*Smoke, error) {
+	s := &Smoke{
+		logger: logger,
+	}
+
+	if s.logger == nil {
+		s.logger = log.NewNopLogger()
+	}
+
+	// use the current context in kubeconfig. this falls back to in-cluster config if kubeconfig is empty
+	config, err := clientcmd.BuildConfigFromFlags("", opts.Kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// add default tasks
+	// TODO: need to add a deletePodTask for cluster pods,
+	// TODO: currently script generates random number and appends it
+	// TODO: to the pod name. Should do this with a label selector,
+	// TODO: by adding a deletePodByLabelTask or such
+	s.tasks = append(s.tasks,
+		repeatingTask{
+			Task: &deletePodTask{
+				logger:    s.logger,
+				clientset: clientset,
+				namespace: opts.Namespace,
+				pod:       "grafana-agent-0",
+			},
+			frequency: opts.ChaosFrequency,
+		},
+		repeatingTask{
+			Task: &scaleDeploymentTask{
+				logger:      s.logger,
```

Comment:
very light suggestion: you can save yourself on typing here by making this 

```go
logger: log.With(s.logger, "task", "scale_deployment", "deployment", "avalanche") 
```

then you would be able to avoid all the formatting of the log messages and have something more generic: 

```go
// logs as ts=<time> level=error task=scale_deployment deployment=avalanche msg="failed to get autoscalingv1.Scale" err=<error> 
level.Error(t.logger).Log("msg", "failed to get autoscalingv1.Scale", "err", err)

```

Reply:
nice :+1: this is clean